### ISSUE TITLE
[typescript-angular2] Template literals on url path does not match variable names

### ISSUE THREAD

**Issue by soycabanillas** [swagger-api/swagger-codegen#4745]
Generating this swagger definition:

  '/v2.01/Users/{UserId}':
    get:
      operationId: UsersGet
      parameters:
        - name: UserId
          in: path
          required: true
          type: integer
          format: int64

the typescript-angular2 code that is created is:

    public usersGetWithHttpInfo(userId: number, extraHttpRequestParams?: any): Observable<Response> {
        const path = this.basePath + `/v2.01/Users/${UserId}`;
        ...

As you can see, the variable name is 'userId', but the const Path is using 'UserId'.
So an error arises trying to compile the code.

I've used the option modelPropertyNaming with value 'original', but it does not affect the basePath.

Tried using a request to the swagger online code generator through:

    curl -k -X POST -H "content-type:application/json" -d '{"options":{"modelPropertyNaming":"original"},"swaggerUrl":"https://XXX/swagger.json"}' https://generator.swagger.io/api/gen/clients/typescript-angular2

**Comment by wing328** [comment]
@soycabanillas thanks for reporting the issue. To fix the issue, we need to start with the following line:

https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/typescript-angular2/api.mustache#L64

and apply similar fix found in typescript-angular: https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/typescript-angular/api.mustache#L32

May I know if you've time to contribute a fix?

**Comment by soycabanillas** [comment]
Tight schedule right now, but I'm interested on fixing this issue because we want to automate the generation of SDKs after every release of our product. I'll work on it on a couple of weeks if still not fixed by somebody else.

**Comment by wing328** [comment]
@soycabanillas it's actually a quick fix and I can do it for you. Would you have time to perform tests afterwards?

**Comment by soycabanillas** [comment]
Of course!

**Comment by wing328** [comment]
I've filed #4756 to fix the issue. Please run the following to pull the fix:
```
git checkout -b wing328-ts-angular2-fix-path-param master
git pull https://github.com/wing328/swagger-codegen.git ts-angular2-fix-path-param
```

I also found that the path variables are explicitly prepended with a dollar sign '$': https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/TypeScriptAngular2ClientCodegen.java#L208, and I do not know the reason behind.

cc @sebastianhaas @Vrolijkx

**Comment by wing328** [comment]
@soycabanillas did you have a chance to perform a test?

**Comment by soycabanillas** [comment]
Sorry for the delay.

Looks good to me:

    public usersGetWithHttpInfo(userId: number, extraHttpRequestParams?: any): Observable<Response> {
        const path = this.basePath + '/v2.01/Users/${UserId}'
                    .replace('${' + 'UserId' + '}', String(userId));
        ...

My workmate will try it soon, as he is the user of the sdk, and I'll give you feedback, but it seems that everything it's ok now.

Beside this, I don't get why one has to set the modelPropertyNaming to original and it's not the default option. Could you point me out where to find the explanation of this design decision or just give me two cents?
I've read on https://github.com/swagger-api/swagger-codegen/issues/2766#issuecomment-228650240 that, ideally, there should be a JSON deserializer for typescript/javascript. And on https://github.com/swagger-api/swagger-codegen/pull/4264 I see that @TiFu is working on it. Will this allow to use any modelPropertyNaming once the pull request is finished?

Thanks!

**Comment by soycabanillas** [comment]
The typescript compiler is complaining about another issue that the repo you provided introduces (it didn't happen with the online codegen):

    public usersGetListWithHttpInfo(page?: number, perPage?: number, extraHttpRequestParams?: any): Observable<Response> {
    ...
        if (page !== undefined) {
            if(page instanceof Date) {
                queryParameters.set('Page', <any>page.d.toISOString());
            } else {
                queryParameters.set('Page', <any>page);
            }
        }    ...

It complains about using instanceof with a number data type (The left-hand side of an 'instanceof' expression must be of type 'any', an object type or a type parameter) and the reference of page.d (Property 'd' does not exist on type 'number & Date').

This is the code the online codegen provided:

        if (page !== undefined) {
            queryParameters.set('Page', <any>page);
        }

I've searched the repository and found that maybe this pull request introduced a bug:
https://github.com/swagger-api/swagger-codegen/pull/3986

**Comment by wing328** [comment]
@gigo1980 would you please take a look at @soycabanillas findings above?

**Comment by gigo1980** [comment]
@wing328: i have create a new pull request that fix that issue #4812

**Comment by wing328** [comment]
@soycabanillas both https://github.com/swagger-api/swagger-codegen/issues/4756 and #4812 have been merged into master. Please pull the latest master to give it a try.

**Comment by larserikgk** [comment]
I was having the exact problem that @soycabanillas described before #4812 was merged. I built a 2.2.3 snapshot from cffaaa027adb2ddb32ed7e29dda3f1f777dff3b3 today and I no longer have this issues @wing328.

**Comment by wing328** [comment]
@larserikgk thanks for doing a test to confirm the fix.
